vmbench.py

```diff
@@ -0,0 +1,472 @@
+
+    pip install --requirement=requirements.txt pandas tabulate
+
+
+Synopsis
+========
+::
+
+    # Start CrateDB with 14 GB heap memory.
+    vmbench.py start
+
+    # Provision database with `uservisits` table, once.
+    vmbench.py setup
+
+    # Invoke list of benchmark specifications, saving corresponding result files
+    # to /home/crate/cratedb-benchmark-results.
+    vmbench.py run
+
+    # Run each scenario five times.
+    vmbench.py run 5
+
+    # Summarize numbers on designated variant.
+    python vmbench.py collect vanilla
+    python vmbench.py collect idlepoll
+    python vmbench.py collect idlepoll-nosmt
+
+    # Compare numbers between variants side by side.
+    python vmbench.py summary
+
+
+Production
+==========
+
+Convenient shortcut to address `vmbench` program within virtualenv::
+
+    vmbench() {
+        sudo --user=crate /opt/crate-benchmarks/.venv/bin/python /opt/crate-benchmarks/vmbench.py "$@"
+    }
+
+::
+
+    # Bootstrap environment.
+    # Note: Run this after each reboot.
+    vmbench start
+    vmbench setup
+
+    # Run benchmarks.
+    # Note: Change Linux kernel parameters and iterate to create "variants".
+    vmbench run 5
+
+    # Compare benchmark results between variants.
+    vmbench summary
+
+
+References
+==========
+
+- https://engineering.mongodb.com/post/reducing-variability-in-performance-tests-on-ec2-setup-and-key-results
+- https://engineering.mongodb.com/post/repeatable-performance-tests-cpu-options-are-best-disabled
+- https://www.kernel.org/doc/html/latest/admin-guide/pm/cpuidle.html
+- https://www.kernel.org/doc/html/latest/admin-guide/hw-vuln/l1tf.html#smt-control
+"""
+import datetime
+import json
+import platform
+import subprocess
+import sys
+import textwrap
+from collections import OrderedDict
+from pathlib import Path
+
+from cr8.run_crate import run_crate
+from cr8.run_spec import run_spec
+
+from compare_results import read_results
+from util import human_readable_byte_size
+
+
+class Scenario:
+    """
+    Run a defined set of `run-spec` files and record their outcomes with different variants.
+    The variants are runtime environment tunings with Linux kernel parameters.
+    """
+
+    CRATEDB_HOST = "localhost:4200"
+
+    specs = [
+        {"file": "specs/queries.toml"},
+        {"file": "specs/select/hyperloglog.toml"},
+        # {"file": "specs/insert_single.py", "full": True},
+        {"file": "specs/insert_bulk.toml", "full": True},
+        {"file": "specs/insert_unnest.py", "full": True},
+    ]
+
+    variants = [
+        "vanilla",
+        "idlepoll",
+        "idlepoll-nosmt",
+    ]
+
+    def __init__(self):
+        # TODO: Optionally use path from `VMBENCH_RESULTS` environment variable.
+        home = Path.home()
+        self.resultfile_path = home / "cratedb-benchmarks-results"
+
+    def start_cratedb(self):
+        # TODO: Make version and heap size configurable.
+        run_crate(version="4.7", env=["CRATE_HEAP_SIZE=14G"], keep_data=True)
```

Comment:
@WalBeh told me that assigning 14G of heap space would be too much. While they used to assign 50% of system memory to CrateDB's heap, they are only assigning 20% these days. That would be 3.2G for a machine with 16G of system memory.

What would be your advice, @mfussenegger?